### PR TITLE
Implemented NCE loss.

### PR DESCRIPTION
**Pull request by k920049**
## Describe your changes
- [v] Implemented NCE loss with log sum exponential techniques.
- [v] Added some test cases.

### CODE REVIEW

File: info_nce/__init__.py [hunk 1]
```diff
@@ -105,17 +105,43 @@ def info_nce(query, positive_key, negative_keys=None, temperature=0.1, reduction
 
         # First index in last dimension are the positive samples
         logits = torch.cat([positive_logit, negative_logits], dim=1)
-        labels = torch.zeros(len(logits), dtype=torch.long, device=query.device)
+
+        # Apply temperature
+        logits = torch.div(logits, temperature)
+
+        # Compute positive terms
+        nominator = -1. * positive_logit.squeeze()
```
Comment (RElbers) [inline info_nce/__init__.py:113]: positive_logit is not scaled by the temperature here.

File: info_nce/__init__.py [hunk 2]
```diff
@@ -105,17 +105,43 @@ def info_nce(query, positive_key, negative_keys=None, temperature=0.1, reduction
...
+
+        # Apply log sum exp tricks here
+        a_max, _ = torch.max(logits, dim=1, keepdim=True)
+        denominator = torch.log(torch.sum(torch.exp(logits - a_max), dim=1)) + a_max.squeeze()
```
Comment (RElbers) [inline info_nce/__init__.py:117]: Pytorch has a function which does this: torch.logsumexp(...)

**Review by RElbers — COMMENTED** [review]
There is a bug where the temperature is not applied to the positive logits. However, your implementation is equivalent to the original if the temperature bug is fixed. 

To try for yourself, make a copy of the original implementation of the `info_nce` function and name it `info_nce_original`. Then run this:

```
batch_size, num_negative, embedding_size = 3, 48, 128
query = torch.randn(batch_size, embedding_size).double()
positive_key = torch.randn(batch_size, embedding_size).double()
negative_keys = torch.randn(num_negative, embedding_size).double()

output0 = info_nce(query, positive_key, negative_keys, temperature=1.0, negative_mode='unpaired', reduction='none')
output1 = info_nce_original(query, positive_key, negative_keys, temperature=1.0, negative_mode='unpaired', reduction='none')
print(torch.allclose(output0, output1))
```

**Comment by k920049** [comment]
After carefully examining the Cross Entropy loss, I realized that the function indeed does not ignore off diagonal components. Thanks for the suggestion and carefully testing the results.
